pkg/controller/servingcert/controller/secret_updating_controller.go

```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+		return nil
 	}
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("unable to get service %s/%s: %v", secret.Namespace, serviceName, err))
-		return false, nil
-	}
-
-	if sharedService.Annotations[api.ServingCertSecretAnnotation] != secret.Name && sharedService.Annotations[api.AlphaServingCertSecretAnnotation] != secret.Name {
-		return false, nil
-	}
-	if secret.Annotations[api.ServiceUIDAnnotation] != string(sharedService.UID) && secret.Annotations[api.AlphaServiceUIDAnnotation] != string(sharedService.UID) {
-		return false, nil
+		utilruntime.HandleError(fmt.Errorf("unable to get service %s/%s: %v", sharedSecret.Namespace, serviceName, err))
+		return nil
 	}
+	return service
+}
 
+func (sc *serviceServingCertUpdateController) requiresRegeneration(service *v1.Service, secret *v1.Secret, minTimeLeft time.Duration) bool {
 	// if we don't have an ownerref, just go ahead and regenerate.  It's easier than writing a
 	// secondary logic flow.
-	if !ocontroller.HasOwnerRef(secret, ownerRef(sharedService)) {
-		return true, sharedService
+	if !ocontroller.HasOwnerRef(secret, ownerRef(service)) {
+		return true
 	}
-
 	// if we don't have the annotation for expiry, just go ahead and regenerate.  It's easier than writing a
 	// secondary logic flow that creates the expiry dates
 	expiryString, ok := secret.Annotations[api.ServingCertExpiryAnnotation]
 	if !ok {
 		expiryString, ok = secret.Annotations[api.AlphaServingCertExpiryAnnotation]
 		if !ok {
-			return true, sharedService
+			return true
 		}
 	}
 	expiry, err := time.Parse(time.RFC3339, expiryString)
 	if err != nil {
-		return true, sharedService
+		return true
 	}
 
 	if time.Now().Add(sc.minTimeLeftForCert).After(expiry) {
-		return true, sharedService
+		return true
 	}
 
-	return false, nil
+	return false
+}
+
+// ensureSecretData will return 2 bools 1) need to update and 2) need to replace with new data
+func (sc *serviceServingCertUpdateController) ensureSecretData(service *v1.Service, secretCopy *v1.Secret) (bool, bool) {
+	update := false
+	replace := false
+	tlsCert, ok := secretCopy.Data[v1.TLSCertKey]
+	tlsKey, ok2 := secretCopy.Data[v1.TLSPrivateKeyKey]
+	if ok && ok2 {
+		// Don't update, replace unless len data is not 2
+		// This takes care of case where someone added an extra data field. Remove all but 2 keys
+		if len(secretCopy.Data) != 2 {
+			// don't regenerate, just get rid of bad data
+			secretCopy.Data = map[string][]byte{
+				v1.TLSCertKey:       tlsCert,
+				v1.TLSPrivateKeyKey: tlsKey,
+			}
+			update = true
+		}
+	} else {
+		update = true
+	}
+	// This ensures someone did not replace secret cert data with ascii art
+	// Check for valid cert data, replace with valid data if not
+	block, _ := pem.Decode([]byte(tlsCert))
+	if block == nil {
+		// Regenerate the secret
+		glog.Infof("Error decoding cert bytes %s from secret: %s namespace: %s, replacing cert", v1.TLSCertKey, secretCopy.Name, secretCopy.Namespace)
+		return true, true
+	} else {
```

Comment:
No need for an `else` as you're returning above, just do it outside any condition scope.

Reply:
ugh again :(  'elses' fixed.